ReactiveCocoaTests/Swift/SignalSpec.swift

```diff
@@ -14,121 +14,771 @@ import ReactiveCocoa
+				
+				signal.observe(next: { number in
+					fromSignal.append(number)
+				}, completed: {
+					completed = true
+				})
+				
+				expect(fromSignal).to(beEmpty())
+				expect(completed).to(beFalsy())
+				
+				sendNext(observer, 1)
+				expect(fromSignal).to(equal([1]))
+				
+				sendNext(observer, 2)
+				expect(fromSignal).to(equal([1, 2]))
+				
+				expect(completed).to(beFalsy())
+				sendCompleted(observer)
+				expect(completed).to(beTruthy())
 			}
 		}
 
 		describe("observe") {
-			pending("should stop forwarding events when disposed") {
+			var testScheduler: TestScheduler!
+			
+			beforeEach {
+				testScheduler = TestScheduler()
+			}
+			
+			it("should stop forwarding events when disposed") {
+				let disposable = SimpleDisposable()
+				
+				let signal: Signal<Int, NoError> = Signal { observer in
+					testScheduler.schedule {
+						for number in [1, 2] {
+							sendNext(observer, number)
+						}
+						sendCompleted(observer)
+						sendNext(observer, 4)
+					}
+					return disposable
+				}
+				
+				var fromSignal: [Int] = []
+				signal.observe(next: { number in
+					fromSignal.append(number)
+				})
+				
+				expect(disposable.disposed).to(beFalsy())
+				expect(fromSignal).to(beEmpty())
+				
+				testScheduler.run()
+				
+				expect(disposable.disposed).to(beTruthy())
+				expect(fromSignal).to(equal([1, 2]))
 			}
 
-			pending("should not trigger side effects") {
+			it("should not trigger side effects") {
+				var runCount = 0
+				let signal: Signal<(), NoError> = Signal { observer in
+					runCount += 1
+					return nil
+				}
+				
+				expect(runCount).to(equal(1))
+				
+				signal.observe()
+				expect(runCount).to(equal(1))
 			}
 
-			pending("should release observer after termination") {
+			it("should release observer after termination") {
+				weak var testStr: NSMutableString?
+				let (signal, sink) = Signal<Int, NoError>.pipe()
+
+				let test: () -> () = {
+					var innerStr: NSMutableString = NSMutableString()
+					signal.observe(next: { value in
+						innerStr.appendString("\(value)")
+					})
+					testStr = innerStr
+				}
+				test()
+
+				sendNext(sink, 1)
+				expect(testStr).to(equal("1"))
+				sendNext(sink, 2)
+				expect(testStr).to(equal("12"))
+
+				sendCompleted(sink)
+				expect(testStr).to(beNil())
 			}
 
-			pending("should release observer after disposal") {
+			it("should release observer after disposal") {
+				weak var testStr: NSMutableString?
+				var disposable: Disposable!
+				let signalProducer = SignalProducer<Int, NoError>() { sink, producerDisposable -> () in
+					sendNext(sink, 1)
+					sendNext(sink, 2)
+				}
+
+				let test: () -> () = {
+					var innerStr: NSMutableString = NSMutableString()
+					disposable = signalProducer.start(next: { value in
+						innerStr.appendString("\(value)")
+					})
+					testStr = innerStr
+				}
+				test()
+
+				expect(testStr).to(equal("12"))
+
+				disposable.dispose()
+//				expect(testStr).to(beNil())
```

Comment:
It fails too.
On Mon, Feb 23, 2015 at 1:46 AM Justin Spahr-Summers <
notifications@github.com> wrote:

> In ReactiveCocoaTests/Swift/SignalSpec.swift
> <https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1702#discussion_r25135666>
> :
>
> > +					sendNext(sink, 2)
> > +				}
> > +
> > +				let test: () -> () = {
> > +					var innerStr: NSMutableString = NSMutableString()
> > +					disposable = signalProducer.start(next: { value in
> > +						innerStr.appendString("\(value)")
> > +					})
> > +					testStr = innerStr
> > +				}
> > +				test()
> > +
> > +				expect(testStr).to(equal("12"))
> > +
> > +				disposable.dispose()
> > +//				expect(testStr).to(beNil())
>
> Does it work if you use toEventually instead?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ReactiveCocoa/ReactiveCocoa/pull/1702/files#r25135666>
> .
>